### ISSUE TITLE
ViveknSentimentApproach: Question regarding positive and negative sources

### ISSUE THREAD

**Issue by robmoore** [JohnSnowLabs/spark-nlp#50]
## Description

I've create a Jupyter notebook using the ViveknSentimentApproach along the lines of the notebook provided in the [docs](http://nlp.johnsnowlabs.com/notebooks.html). Rather than using previously created files for the positive and negative training data, I am generating the files within the notebook itself. When I write the training sets to the filesystem (ie, `df.write.mode('overwrite').text("train-data-positive")`, I end up with multiple data files (one per partition) and a number of metadata files (crc files and a `_SUCCESS` file). 

For example:

```
./_SUCCESS
./._SUCCESS.crc
./part-00000-1c1f1cff-5720-4088-80d7-6c15e0fe9bd6-c000.txt
./.part-00000-1c1f1cff-5720-4088-80d7-6c15e0fe9bd6-c000.txt.crc
...
```
If I then reference the directory where these files are written out (eg, `train-data-positive` and `train-data-negative`), I get back an error with the following stacktrace:

```
Py4JJavaError: An error occurred while calling o1665.fit.
: java.io.FileNotFoundException: Invalid file path
	at java.io.FileInputStream.<init>(FileInputStream.java:133)
	at scala.io.Source$.fromFile(Source.scala:91)
	at scala.io.Source$.fromFile(Source.scala:76)
	at scala.io.Source$.fromFile(Source.scala:54)
	at scala.io.Source$.fromFile(Source.scala:60)
	at com.johnsnowlabs.nlp.util.io.ResourceHelper$SourceStream$$anonfun$2.apply(ResourceHelper.scala:35)
	at com.johnsnowlabs.nlp.util.io.ResourceHelper$SourceStream$$anonfun$2.apply(ResourceHelper.scala:35)
	at scala.Option.getOrElse(Option.scala:121)
	at com.johnsnowlabs.nlp.util.io.ResourceHelper$SourceStream.<init>(ResourceHelper.scala:35)
	at com.johnsnowlabs.nlp.util.io.ResourceHelper$.wordCount(ResourceHelper.scala:192)
	at com.johnsnowlabs.nlp.util.io.ResourceHelper$$anonfun$wordCount$2.apply(ResourceHelper.scala:196)
	at com.johnsnowlabs.nlp.util.io.ResourceHelper$$anonfun$wordCount$2.apply(ResourceHelper.scala:196)
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:434)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:440)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
	at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:59)
	at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:104)
	at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:48)
	at scala.collection.TraversableOnce$class.to(TraversableOnce.scala:310)
	at scala.collection.AbstractIterator.to(Iterator.scala:1336)
	at scala.collection.TraversableOnce$class.toBuffer(TraversableOnce.scala:302)
	at scala.collection.AbstractIterator.toBuffer(Iterator.scala:1336)
	at scala.collection.TraversableOnce$class.toArray(TraversableOnce.scala:289)
	at scala.collection.AbstractIterator.toArray(Iterator.scala:1336)
	at com.johnsnowlabs.nlp.util.io.ResourceHelper$.wordCount(ResourceHelper.scala:197)
	at com.johnsnowlabs.nlp.util.io.ResourceHelper$$anonfun$wordCount$1.apply(ResourceHelper.scala:189)
	at com.johnsnowlabs.nlp.util.io.ResourceHelper$$anonfun$wordCount$1.apply(ResourceHelper.scala:189)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
	at scala.collection.TraversableLike$$anonfun$flatMap$1.apply(TraversableLike.scala:241)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at scala.collection.TraversableLike$class.flatMap(TraversableLike.scala:241)
	at scala.collection.mutable.ArrayOps$ofRef.flatMap(ArrayOps.scala:186)
	at com.johnsnowlabs.nlp.util.io.ResourceHelper$.wordCount(ResourceHelper.scala:189)
	at com.johnsnowlabs.nlp.annotators.sda.vivekn.ViveknSentimentApproach.train(ViveknSentimentApproach.scala:45)
	at com.johnsnowlabs.nlp.annotators.sda.vivekn.ViveknSentimentApproach.train(ViveknSentimentApproach.scala:14)
	at com.johnsnowlabs.nlp.AnnotatorApproach.fit(AnnotatorApproach.scala:28)
	at sun.reflect.GeneratedMethodAccessor68.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:357)
	at py4j.Gateway.invoke(Gateway.java:280)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:214)
	at java.lang.Thread.run(Thread.java:748)
```
However, if I delete the `_SUCCESS` and crc files then all is well.

## Expected Behavior

I can modify my notebook to delete these files but it seems like the code should filter on `*.txt` if those are the only acceptable file names.

**Comment by saifjsl** [comment]
Thank you @robmoore 
Certainly, the ResourceHelper needs a lot of work to improve it to allow different formats, and also throw proper errors when format or file types are not allowed.

I will take a look to this specific case and see what can be done.

**Comment by robmoore** [comment]
@saifjsl Thanks! I really appreciate the work the team has done on SparkNLP and I have enjoyed working with it so my apologies if my ticket came off as entitled. My intention was to share the friction point I ran into while working on my [notebook](https://github.com/robmoore/jupyter-trump-sentiment) as a way to learn more about SparkNLP.

**Comment by robmoore** [comment]
I've reworked my notebook to use the directory approach after removing all of the metadata files mentioned above. I don't get an error but it's clear that the annotator isn't working (all my sentiments are negative). Is there some assumption about what the filenames will be containing the data?

**Comment by saifjsl** [comment]
ViveknSentimentAnalysis will take two inputs, a folder containing positive .txt files and a folder containing negative.txt. Then, it will learn from those. You can take a look both at the python notebook examples in python/example folder or at the unit tests of such annotator, to see a use case example. Example training data is also included in the library

**Comment by robmoore** [comment]
Thanks. I had consulted the example provided in this project previously. It sounds like what I've undertaken in the approach I took subsequent to asking this question is what you outline. 

I do the following in my notebook:
```
# Delete _SUCCESS and .crc files since the ViveknSentimentApproach chokes on them
def delete_metadata(dirname):
    # Use Hadoop library directly to work to filesystem
    Path = sc._gateway.jvm.org.apache.hadoop.fs.Path
    FileSystem = sc._gateway.jvm.org.apache.hadoop.fs.FileSystem
    Configuration = sc._gateway.jvm.org.apache.hadoop.conf.Configuration

    fs = FileSystem.get(Configuration())
    fs.delete(Path(dirname + "/_SUCCESS"), False);
    files = fs.listStatus(Path(dirname))
    for file in files:
        fs.delete(Path(dirname, "." + file.getPath().getName() + ".crc"), False)

positive_data = train_data.filter(train_data.Class == "1").select("norm_text")
positive_data.write.mode('overwrite').text("train-data-pos")
delete_metadata("train-data-pos")

negative_data = labeled_data.filter(train_data.Class == "0").select("norm_text")
negative_data.write.mode('overwrite').text("train-data-neg")
delete_metadata("train-data-neg")
...
sentiment_detector = ViveknSentimentApproach().setInputCols(["spell", "sentence"]) \
    .setOutputCol("sentiment").setPositiveSource("train-data-pos") \
    .setNegativeSource("train-data-neg").setPruneCorpus(False)   
```
As I said above, I do not get any error messages using this approach but my results are all negative so I suspect something is not working. Is there any logging I could enable to provide more information?

**Comment by saifjsl** [comment]
Thanks @robmoore . If you can provide me the contents of train-data-pos and train-data-neg as well as a piece of your target data, I may be able to help you and identify where the issue comes from.

For starters, if you are able to make it work with provided resources located in:
/src/test/resources/sentiment.parquet
using positive and negative content examples from:
/src/test/resources/vivekn/positive~negative
You can confirm whether this may be something with your inputs or with the pipeline.

Also, are you sure the files you are writing are a single partition of a .txt file? otherwise, as I suggested before, we are only able to read txt files or folders containing txt files written in disk (not from hdfs).

**Comment by robmoore** [comment]
Thanks @saifjsl. I was able to read in the vivekn pos/neg files successfully and received something other than all negative sentiments using the same pipeline as before. 

I compared the input I'm using to the vivekn version and the only difference that stood out to me is that the last line of the vivekn version doesn't end in a linefeed whereas mine does.

Perhaps you can spot something I'm missing. You can look at the files I'm using here: https://www.dropbox.com/s/s46smth5wg69had/spark-nlp-50.tar.xz?dl=0

**Comment by GreGGus** [comment]
Same issue in Spark/Scala

HDFS : 
    -/user/me/tweet/txt/neg/neg_tweet.txt
    -/user/me/tweet/txt/pos/pos_tweet.txt

`    val sentiment_analysis = new ViveknSentimentApproach()
      .setInputCols(Array("token", "sentence"))
      .setOutputCol("vivekn")
      .setPositiveSourcePath("/user/me/tweet/txt/pos")
      .setNegativeSourcePath("/user/me/tweet/txt/neg")
`

Doesn't work.

`    val sentiment_analysis = new ViveknSentimentApproach()
      .setInputCols(Array("token", "sentence"))
      .setOutputCol("vivekn")
      .setPositiveSourcePath("/user/me/tweet/txt/neg/neg_tweet.txt")
      .setNegativeSourcePath("/user/me/tweet/txt/pos/pos_tweet.txt")
`

Doesn't work.

Why did I miss ?

**Comment by GreGGus** [comment]
My bad

> we are only able to read txt files or folders containing txt files written in disk (not from hdfs).

**Comment by saifjsl** [comment]
Hi @Griggs,  next release is reworking all external inputs reader to allow filesystem or hdfs just as spark does.

**Comment by saifjsl** [comment]
I mean 1.4.0

**Comment by saifjsl** [comment]
Should be fixed in version #1.5.0
